BUG: ma.put expands nomask

jjhelmus opened the pull request:
Previously when put was used on a MaskedArray with nomask the mask would be
incorrectly set to the mask of the values argument.

closes #6425

jjhelmus committed:
BUG: ma.put expands nomask

Previously when put was used on a MaskedArray with nomask the mask would be
incorrectly set to the mask of the values argument.

closes #6425

ahaldane commented:
Looks good. Some little comments:

 * Should you add a clause for the case that both `self` and `values` have a mask of `nomask`, to leave the mask alone? This way we avoid creating a full boolean mask, then shrinking it at the end. Or is that uneccesary optimization? If there is an uncomplicated way to reorganize the if statements like that, I'd be for it.
 * Note that the fact that we make a copy of the mask is a bit weird: Consider a MaskedArray which is actually a slice of another MaskedArray (its data is a view of the other's). `put`ting to the view will update the original array's data, but not its mask. This issue was already discussed in #5580, though, and it turns out the documentation states that masks of views are copies, and also that this is currently unfixable. So we can leave it alone here too, just wanted to make anyone reading aware.

jjhelmus commented:
I agree that creating a full boolean mask and then shrinking it is undesirable and adding logic to address this case is not too difficult.  I was not aware of the issue with views of MaskedArrays but will keep it in mind. 

I refactored the method to add a short circuit when both `self` and `values` have a mask of `nomask`, separated the setting of the data and mask and simplified the masking logic.  The method now has two return locations which may be undesired.  If a single return is desired the masking logic can be rewritten as:

``` Python
        if self._mask is not nomask or getmask(values) is not nomask:
            m = getmaskarray(self).copy()
            if getmask(values) is nomask:
                m.put(indices, False, mode=mode)
            else:
                m.put(indices, values._mask, mode=mode)
            m = make_mask(m, copy=False, shrink=True)
            self._mask = m
```

ahaldane commented on numpy/ma/tests/test_core.py:
```diff
@@ -2671,6 +2671,24 @@ def test_put(self):
+        self.assertTrue(x[1] is not masked)
+        self.assertTrue(x[2] is not masked)
+        self.assertTrue(x[3] is not masked)
+        self.assertTrue(z[1] is masked)
```
Are these 4 asserts needed? If they are only testing basic MA functionality which is tested elsewhere, I favor removing them for clarity.

The asserts below are good and make clear what we are testing.

ahaldane commented:
That looks good to me now, @jjhelmus. After the question of the asserts, I will be happy to merge.

Could you also squash everything into a single commit, using `git rebase`? I think the fixups are small enough they don't require separate commits.

jjhelmus commented:
The first four assert statements were unnecessary, I removed them and rebased into a single commit.

ahaldane commented:
Great! I'll go ahead and merge. Thanks @jjhelmus
